### PR TITLE
ARROW-8427: [C++][Dataset] Only apply ignore_prefixes to selector results

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

File: cpp/src/arrow/dataset/discovery.h [hunk 1]
```diff
@@ -165,15 +165,17 @@ struct FileSystemFactoryOptions {
   // (resulting in an error at scan time).
   bool exclude_invalid_files = false;
 
-  // Files matching one of the following prefix will be ignored by the
-  // discovery process. This is matched to the basename of a path.
+  // Files and directories matching one of the following prefixes will be ignored
```
Comment (fsaintjacques) [inline cpp/src/arrow/dataset/discovery.h:168]: I'd make `When using a selector (and not to an explicit list of files),` the first phrase so it doesn't catch off-guard those who don't read until the end.

Maybe rename the property name, e.g. `selector_ignore_prefixes`?

**Review by fsaintjacques — CHANGES_REQUESTED** [review]
Is this already covered by a unit test? If not, add both the inclusive and exclusive part, e.g. selector with ignored file, and manual list with a path that matches one of the prefixed but is not ignored.

File: cpp/src/arrow/dataset/discovery.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/discovery.h:168]: Good call, will reword

**Comment by jorisvandenbossche** [comment]
I am adding a python test (also to confirm it is consistent with the old ParquetDataset implementation)

**Comment by jorisvandenbossche** [comment]
@github-actions crossbow submit conda-win-vs2015-py37

**Comment by jorisvandenbossche** [comment]
Hmm, the test I am writing is also passing on master locally ... :-(

For me, it actually *keeps* the file even if one of the parent directories has an underscore .. (it only prunes the directories)

**Review by kszucs — APPROVED** [review]
LGTM

**Comment by kszucs** [comment]
I'll let @fsaintjacques make the final call.

**Pull request closed without merging**